push/ups_integration_with_openshift.md

```diff
@@ -0,0 +1,100 @@
+In UPS, the developer creates a 'Push Application' and 1 or more 'Variants' depending on the number & type of Apps they are developing outside of Openshift.
+The developer also has to somehow get the Push configuration from UPS and into their App.
+
+The fourth problem is allowing a developer to write custom logic that dynamically sends Push messages to App users.
+This can be done by calling the UPS REST API with the correct creentials/API Key.
+However, configuration is required in UPS to get these credentials.
+These credentials then need to be managed and exposed to the custom code.
+
+All of these problems highlight pain points in the developer experience.
+These problems don't block the developer, but do make it harder than it could be to add Push integration to their Apps. 
+
+## Terms
+
+- TODO
+
+## OAuth provider support 
+
+To address the first & second problem, UPS will be modified to support different auth providers.
+The 2 providers will be Keycloak (to maintain backwards compatibility for existing users) and a generic OAuth Provider.
+The OAuth flow will follow the same pattern & flow used by the [OpenShift OAuth Proxy](https://github.com/openshift/oauth-proxy).
+The UPS APB will need to be modified to include an OAuth ServiceAccount, similar to how [Jenkins uses a ServiceAccount](https://github.com/openshift/jenkins-openshift-login-plugin#browser-access).
+UPS will use this ServiceAccount as an OAuth Client to verify the User has the necessary permissions in OpenShift to use UPS.
+The permissions will be:
+* edit access on the namespace in OpenShift => Standard Developer Account in UPS
+* admin access on the namespace in OpenShift => Admin Account in UPS
+
+These changes will mean UPS will no longer have a dependency on Keycloak when running in OpenShift.
+Minimal changes will be required to the UPS APB.
+Also, the developer doesn't have to know about or be concerned with where or how the auth provider is setup and configured for UPS.
+
+## OpenShift Integration
+
+To address the third problem, a sidecar will be added to the UPS APB.
+This sidecar will provide background bi-directional syncing of resources and metadata between OpenShift and UPS.
+The sidecar will be written in Go and utilise a kubernetes client for watching & updating Kubernetes resources.
+It will use a SerivceAccount for auth when watching resources.
+The sidecar will also use a ServiceAccount to access the UPS Admin REST API for reading & updating UPS resources.
+
+### Syncing
+
+Syncing will happen *from* OpenShift *to* UPS for the following resources:
+
+* MobileClient => Variant
+* Secret (p12 or secret key) => Variant configuration
+
+The Secrets will have an appropriate label so it can be easily filtered when looking for resources to sync.
+The sidecar will overwrite the Variant fields and configuration to match what the values are in OpenShift.
+This means that any changes to a Variant directly in the UPS UI will be overwritten on next sync.
```

Comment:
this will happen _only_ on the tight Openshift integration

Reply:
Correct. You'll only have the sidecar if you're running on OpenShift, so this will only happen then

Reply:
I think we need to be careful here, as this means a developer will have to interact with OpenShift object in order to change things  in UPS. This might not be the developer experience we want. 
Perhaps a better approach may be something like:
- if the variant does not exist then it is created in UPS and reflected in the CRD with an the variant ID
- if it does exist the CRD is updated with any details present in UPS
- if there is and ID present in the CRD that is no longer present in UPS then it has been deleted in UPS and should be removed from the CRD

This means if developers

Reply:
@maleck13 Here's info on how the syncing works for jenkins (which may influence our thinking) https://github.com/openshift/jenkins-sync-plugin#openshift-jenkins-sync-plugin
It doesn't look like it's very destructive, but I'd like to try it out and poke around more (chaos monkey)

The only think that get's synced back to OpenShift in that case is Job Runs as Build Objects

> changes in a Jenkins Build Run thats associated with a Jenkins Job gets replicated to an OpenShift Build object (which is created if necessary if the build was triggered via Jenkins)

Reply:
Jira task that will look to understand and define this section better https://issues.jboss.org/browse/AEROGEAR-2326

Reply:
Users will need to pick one of the variants. Sync logic could be complex and can lead to the bugs. I will suggest to let users to specify their variants.